Building with docker can prevent appsignal from downloading proper archive

#### Setup

I am running MacOS. I am using docker to build my application for deployment.

#### Replication Instructions

If I run 

```
mix deps.get
mix deps.compile
```

on my MacOS machine, it will download and compile everything appsignal needs which works perfectly. If I then attempt to do the same within a Dockerfile

```
FROM elixir:1.5.1

RUN apt-get clean && apt-get update -qq && apt-get -y install build-essential curl

COPY ./ /app

WORKDIR /app

RUN mix local.hex --force
RUN mix local.rebar --force
RUN mix deps.get
RUN mix deps.compile
```

I will get a compilation error as follows.

```
/app/_build/prod/lib/appsignal/priv/libappsignal.a: member /app/_build/prod/lib/appsignal/priv/libappsignal.a(__.SYMDEF) in archive is not an object
collect2: error: ld returned 1 exit status
make: *** [all] Error 1
Makefile:22: recipe for target 'all' failed
==> appsignal
could not compile dependency :appsignal, "mix compile" failed. You can recompile this dependency with "mix deps.compile appsignal", update it with "mix deps.update appsignal" or clean it with "mix deps.clean appsignal"
** (Mix) Could not run `make`. Please check if `make` and either `clang` or `gcc` are installed
```

It appears as though the compilation scripts have found that there is already an archive (in this case, of the incorrect architecture) and will not download a new one for the proper architecture. Or maybe it gets unpacked into an architecture agnostic directory?

#### Workaround

You can add `RUN mix deps.clean appsignal` before the `RUN mix deps.get` line and it will work.

You could also create a `.dockerignore` file and add `deps` and `_build` to it.

--- a/.changesets/improve-error-message-on-extension-load-failure.md
+++ b/.changesets/improve-error-message-on-extension-load-failure.md
@@ -1,0 +1,6 @@
+---
+bump: "patch"
+type: "fix"
+---
+
+Improve the error message on extension load failure. The error message will now print more details about the installed and expected architecture when they mismatch. This is most common on apps mounted on a container after first being installed on the host with a different architecture than the container.

--- a/config/config.exs
+++ b/config/config.exs
@@ -8,6 +8,7 @@ if Mix.env() in [:bench, :test, :test_no_nif] do
   config :appsignal, inet: FakeInet
   config :appsignal, io: FakeIO
   config :appsignal, file: FakeFile
+  config :appsignal, os_internal: FakeOS
 
   config :appsignal, appsignal_tracer_nif: Appsignal.Test.Nif
   config :appsignal, appsignal_tracer: Appsignal.Test.Tracer

--- a/dialyzer.ignore-warnings
+++ b/dialyzer.ignore-warnings
@@ -2,4 +2,4 @@ lib/appsignal/json.ex:31: Unknown function 'Elixir.Jason':encode/1
 lib/appsignal/json.ex:32: Unknown function 'Elixir.Jason':'encode!'/1
 lib/appsignal/json.ex:33: Unknown function 'Elixir.Jason':decode/1
 lib/appsignal/json.ex:34: Unknown function 'Elixir.Jason':'decode!'/1
-lib/appsignal.ex:46: The inferred return type of config_change/3 (pid()) has nothing in common with 'ok', which is the expected return type for the callback of the 'Elixir.Application' behaviour
+lib/appsignal.ex:50: The inferred return type of config_change/3 (pid()) has nothing in common with 'ok', which is the expected return type for the callback of the 'Elixir.Application' behaviour

--- a/lib/appsignal.ex
+++ b/lib/appsignal.ex
@@ -8,6 +8,10 @@ defmodule Appsignal do
   helper functions for sending metrics to AppSignal.
   """
 
+  require Mix.Appsignal.Utils
+
+  @os Mix.Appsignal.Utils.compile_env(:appsignal, :os_internal, :os)
+
   use Application
   alias Appsignal.Config
   require Logger
@@ -68,11 +72,7 @@ defmodule Appsignal do
         if Appsignal.Nif.loaded?() do
           Appsignal.Logger.debug("AppSignal started.")
         else
-          Logger.error(
-            "Failed to start AppSignal. Please run the diagnose task " <>
-              "(https://docs.appsignal.com/elixir/command-line/diagnose.html) " <>
-              "to debug your installation."
-          )
+          log_nif_loading_error()
         end
 
       {{:error, :invalid_config}, true} ->
@@ -142,4 +142,56 @@ defmodule Appsignal do
   defdelegate send_error(exception, stacktrace), to: Appsignal.Instrumentation
   defdelegate send_error(kind, reason, stacktrace), to: Appsignal.Instrumentation
   defdelegate send_error(kind, reason, stacktrace, fun), to: Appsignal.Instrumentation
+
+  defp log_nif_loading_error do
+    arch = parse_architecture(to_string(:erlang.system_info(:system_architecture)))
+    {_, target_list} = @os.type()
+    target = to_string(target_list)
+    {install_arch, install_target} = fetch_installed_architecture_target()
+
+    if arch == install_arch && target == install_target do
+      Appsignal.Logger.error(
+        "AppSignal failed to load the extension. Please run the diagnose tool and email us at support@appsignal.com: https://docs.appsignal.com/elixir/command-line/diagnose.html\n",
+        stderr: true
+      )
+    else
+      Appsignal.Logger.error(
+        "The AppSignal NIF was installed for architecture '#{install_arch}-#{install_target}', but the current architecture is '#{arch}-#{target}'. Please reinstall the AppSignal package on the host the app is started: mix deps.compile appsignal --force",
+        stderr: true
+      )
+    end
+  end
+
+  # Parse install report and fetch the architecture and target
+  defp fetch_installed_architecture_target do
+    case File.read(Path.join([:code.priv_dir(:appsignal), "install.report"])) do
+      {:ok, raw_report} ->
+        case Appsignal.Json.decode(raw_report) do
+          {:ok, report} ->
+            %{"build" => %{"architecture" => arch, "target" => target}} = report
+            {parse_architecture(arch), target}
+
+          {:error, reason} ->
+            Appsignal.Logger.error(
+              "Failed to parse the AppSignal 'install.report' file: #{inspect(reason)}",
+              stderr: true
+            )
+
+            {"unknown", "unknown"}
+        end
+
+      {:error, reason} ->
+        Appsignal.Logger.error(
+          "Failed to read the AppSignal 'install.report' file: #{inspect(reason)}",
+          stderr: true
+        )
+
+        {"unknown", "unknown"}
+    end
+  end
+
+  # Transform `aarch64-apple-darwin21.3.0` to `aarch64`
+  defp parse_architecture(arch_parts) do
+    List.first(String.split(arch_parts, "-", parts: 2))
+  end
 end

--- a/mix_helpers.exs
+++ b/mix_helpers.exs
@@ -391,7 +391,7 @@ defmodule Mix.Appsignal.Helper do
     defp map_arch('arm-' ++ _, platform), do: build_for("aarch64", platform)
   end
 
-  defp map_arch(arch, platform), do: {:error, {:unknown, {arch, platform}}}
+  defp map_arch(arch, platform), do: {:error, {:unknown, {to_string(arch), platform}}}
 
   defp build_for(bit, platform) do
     arch = {bit, platform}

--- a/test/appsignal/appsignal_test.exs
+++ b/test/appsignal/appsignal_test.exs
@@ -7,8 +7,78 @@ defmodule AppsignalTest do
     {:ok, _} = start_supervised(Appsignal.Test.Nif)
     {:ok, _} = start_supervised(Appsignal.Test.Span)
     {:ok, _} = start_supervised(Appsignal.Test.Monitor)
+    {:ok, _} = start_supervised(FakeIO)
 
-    :ok
+    [fake_os: start_supervised!(FakeOS)]
+  end
+
+  describe "initialize" do
+    @tag :skip_env_test
+    test "prints warning about extension not being loaded" do
+      Appsignal.initialize()
+      [{:stderr, device_output}] = FakeIO.all()
+      output = String.trim(device_output)
+
+      assert output =
+               "[appsignal][ERROR] AppSignal failed to load the extension. Please run the diagnose tool and email us at support@appsignal.com: https://docs.appsignal.com/elixir/command-line/diagnose.html"
+    end
+
+    @tag :skip_env_test
+    test "prints warning about mismatch extension architecture", %{
+      fake_os: fake_os
+    } do
+      # Make current OS "unknown" value so it will always fail
+      FakeOS.update(fake_os, :type, {:unknown, :unknown})
+      Appsignal.initialize()
+      [{:stderr, device_output}] = FakeIO.all()
+      output = String.trim(device_output)
+
+      assert output =~
+               ~r{AppSignal NIF was installed for architecture '\w+-\w+', but the current architecture is '\w+-unknown'}
+    end
+  end
+
+  describe "initialize when the JSON file not valid" do
+    setup do
+      install_report = Path.join([:code.priv_dir(:appsignal), "install.report"])
+      install_contents = File.read!(install_report)
+      File.write(install_report, "invalid install report JSON")
+
+      on_exit(:reset, fn ->
+        File.write!(install_report, install_contents)
+      end)
+    end
+
+    @tag :skip_env_test
+    test "logs an error that the report file is invalid" do
+      Appsignal.initialize()
+
+      [{:stderr, _architecture_mismatch_error}, {:stderr, report_error}] = FakeIO.all()
+
+      assert report_error =~
+               ~r{Failed to parse the AppSignal 'install.report' file:}
+    end
+  end
+
+  describe "initialize when the report file is not readable" do
+    setup do
+      install_report = Path.join([:code.priv_dir(:appsignal), "install.report"])
+      File.chmod(install_report, 0o000)
+
+      on_exit(:reset, fn ->
+        File.chmod(install_report, 0o644)
+      end)
+    end
+
+    @tag :skip_env_test
+    test "logs an error when the report file is not found" do
+      Appsignal.initialize()
+
+      [{:stderr, _architecture_mismatch_error}, {:stderr, report_error}] = FakeIO.all()
+
+      assert report_error =~
+               ~r{Failed to read the AppSignal 'install.report' file:}
+    end
   end
 
   test "set gauge" do

--- a/test/support/fake_os.ex
+++ b/test/support/fake_os.ex
@@ -1,5 +1,12 @@
 defmodule FakeOS do
   use TestAgent, %{type: {:unix, :linux}}
 
-  def type, do: get(__MODULE__, :type)
+  def type do
+    if alive?() do
+      get(__MODULE__, :type)
+    else
+      # Fall back on original implementation if the fake process is not alive
+      :os.type()
+    end
+  end
 end

